# Solving a System of NonLinear Equations using Newton's Method

## Jennefer Maldonado

(a) Ground State of a Particle

$\frac{x}{\tan(x)} = -y$

$x^2+y^2=s^2$

Where $s=3.5$

(b) Excited State of Particle

$\frac{1}{\tan(x)} - \frac{1}{x^2} = \frac{1}{y} + \frac{1}{y^2}$

$x^2+y^2=s^2$


In [17]:
"""
Author : Jennefer Maldonado
Date   : 3/5/2021
Purpose: Solve Systems of NonLinear Equations
"""

import numpy as np
import math
from mpmath import *

def ground(x,y,s):
    x1 = (x/math.tan(x)) + y
    x2 = x**2 + y**2 - s**2
    return [x1,x2]

def ground_jacobian(x,y,s):
    x1 = (math.tan(x)-x*simplify(sec(x))**2)/(math.tan(x)**2)
    x2 = -1
    x3 = 2*x
    x4 = 2*y
    return [[x1, x2], [x3, x4]]

def g_newtons(x,s):
    f = ground(x[0],x[1],s)
    neg_f = np.negative(f)
    Jf = np.array(ground_jacobian(x[0],x[1],s), dtype='float')
    sk = np.linalg.solve(Jf, neg_f)
    return x + sk

s = 3.5
x1 = 2
y1 = 2
for i in range(0,30):
    x = [x1,y1]
    newpt = g_newtons(x,s)
    print(newpt)
    x1 = newpt[0]
    y1 = newpt[1]


[2.01182212 3.05067788]
[3.015926   2.20755043]
[2.90876071 1.96459259]
[2.72951587 2.21203519]
[2.50751761 2.46486547]
[2.38436728 2.56718279]
[2.3939239  2.55331393]
[2.38858982 2.55825948]
[2.39043721 2.55652427]
[2.38977375 2.55714338]
[2.39000861 2.55692372]
[2.38992505 2.55700182]
[2.38995473 2.55697407]
[2.38994418 2.55698393]
[2.38994793 2.55698043]
[2.38994659 2.55698167]
[2.38994707 2.55698123]
[2.3899469  2.55698139]
[2.38994696 2.55698133]
[2.38994694 2.55698135]
[2.38994695 2.55698134]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]
[2.38994694 2.55698135]


### Analysis for Ground State System
This system solved by Newton's method with 30 iterations. With some trial and error I ended up with [2,2] as my starting point. This causes the system to coverge to approximately [2.38994694, 2.55698135]. When I was trying different intervals like [1,0], Newton's method would fail to converge and oscillate between large numbers.

In [21]:
def excited(x,y,s):
    x1 = (1/x*math.tan(x)) - (1/x**2) - (1/y) - (1/y**2)
    x2 = x**2 + y**2 - s**2
    return [x1,x2]

def excited_jacobian(x,y,s):
    x1 = -1*(math.tan(x)+x*simplify(sec(x))**2)/((x**2)*(math.tan(x)**2)) + 2*(1/x**3)
    x2 = (1/y**2) + (1/y**3)
    x3 = 2*x
    x4 = 2*y
    return [[x1, x2], [x3, x4]]

def e_newtons(x,s):
    f = ground(x[0],x[1],s)
    neg_f = np.negative(f)
    Jf = np.array(excited_jacobian(x[0],x[1],s), dtype='float')
    sk = np.linalg.solve(Jf, neg_f)
    return x + sk

s = 3.5
x1 = 2
y1 = 2
for i in range(0,100):
    x = [x1,y1]
    newpt = e_newtons(x,s)
    print(newpt)
    x1 = newpt[0]
    y1 = newpt[1]

[4.41036289 0.65213711]
[3.60088807 0.27918775]
[3.50076331 0.14797746]
[3.49800352 0.1212222 ]
[3.49842401 0.1061044 ]
[3.49870216 0.09585583]
[3.49889572 0.08824256]
[3.49903807 0.0822694 ]
[3.49914729 0.07740718]
[3.49923384 0.07334211]
[3.49930418 0.06987364]
[3.49936251 0.06686631]
[3.49941169 0.06422463]
[3.49945374 0.061879  ]
[3.49949012 0.05977724]
[3.49952191 0.05787929]
[3.49954993 0.05615386]
[3.49957482 0.05457601]
[3.49959709 0.05312565]
[3.49961712 0.05178631]
[3.49963524 0.0505444 ]
[3.49965171 0.04938856]
[3.49966675 0.04830921]
[3.49968054 0.04729821]
[3.49969323 0.04634858]
[3.49970495 0.04545432]
[3.49971579 0.04461022]
[3.49972587 0.04381172]
[3.49973525 0.04305485]
[3.49974402 0.04233609]
[3.49975221 0.04165232]
[3.4997599  0.04100078]
[3.49976712 0.04037901]
[3.49977392 0.0397848 ]
[3.49978034 0.03921616]
[3.4997864  0.03867129]
[3.49979213 0.03814859]
[3.49979756 0.03764659]
[3.49980272 0.03716394]
[3.49980762 0.03669943]
[3.49981228 0.03625195]
[3.49981672 0.03

### Analysis for Excited State System
This system solved by Newton's method with 30 iterations. With some trial and error I ended up with [2,2] as my starting point same as before. This causes the system to coverge to approximately [3.49992273, 0.02325703]. When I was trying different intervals like [1,0], Newton's method would fail to converge and oscillate between large numbers. Similarly like before Newton's was able to solve the system once the correct interval was given. The downside to this is newton's will fail to converge anywhere outside of this region.